### DISCUSSION TITLE
is it possible to route jackd through roc?

### DISCUSSION THREAD

**Discussion by jave** [roc-streaming/roc-toolkit#252] [Q&A] [answered]
I tried setting up a server with roc using pulseaudio. I can play sound using pulseaudio fine, and receive it on a client machine.

But is it possible to use jackd as well?

I tried this unsucessfully:
- starting jackd with qjackctl, with a dummy driver
- loading module-jack-sink into the pulseaudio
- starting hydrogen, routing through the jackd

this didnt give me any audio on the client though.

**Comment by gavv** [comment]
Hi,

I've no experience with JACK so far, and Roc currently has no direct JACK support. Adding it would be great, but for now nobody is working on it. Pull requests are welcome :)

Theoretically, if you want JACK on the receiver side, using module-jack-sink should work, except that it will cost you additional latency introduced by PulseAudio. You should do the following:

* load module-jack-sink
* ensure that it works; for example, you can run something like `paplay -d <sink_name> test.wav`, where `sink_name` is the name of the sink created by module-jack-sink (find it using `pactl list sinks`)
* load module-roc-sink-input
* connect Roc sink-input to JACK sink, like: `pactl move-sink-input <sink_input_index> <sink_index>`, where `sink_input_index` is the numeric index of the sink-input created by module-roc-sink-input (find it using `pactl list sink-inputs`) and `sink_index` is the numeric index of the sink created by module-jack-sink (find it using `pactl list sinks`)

The last can be also done via pavucontrol GUI.

If it does not work, it is probably a bug in Roc sink-input because theoretically JACK sink is a usual sink, Roc sink-input is a usual sink-input, and everything should just work.

You can also try to use roc-recv command-line tool instead of module-roc-sink-input PulseAudio module. You can run it like this:

```
roc-recv -vv -s rtp+rs8m::10001 -r rs8m::10002 -d pulseaudio -o <sink_name>
```

...where `sink_name` is the name of the sink created by module-jack-sink (find it using `pactl list sinks`).

**Comment by jave** [comment]
Thanks.

In my case I want jack on the sender, and roc on the receiver.

for instance like:

on the sender
hydrogen -> jack -> pulse -> roc 

on the receiver
roc -> pulse -> speakers

I tried settnig something up on the server using module-jack-sink/source, but it didnt seem to work very well.

**Comment by gavv** [comment, marked as answer]
Ah, so you need jack + roc on the sender side.

I guess module-jack-sink will not help you because it allows PA to output to JACK, and you need the opposite.

Not sure about module-jack-source. I guess in this case, if you want to use module-roc-sink, you'll also need module-loopback that will transfer sound from module-jack-source to module-roc-sink because it's not possible to connect a source to a sink directly (see here: [1](https://gavv.github.io/articles/pulseaudio-under-the-hood/#constructing-chains), [2](https://gavv.github.io/articles/pulseaudio-under-the-hood/#virtual-devices-and-streams)). Or, alternatively, you can use roc-send tool instead of module-roc-sink. The tool can grab sound directly from PulseAudio source, which you can specify using `-i` option.

You can also try to do it without PulseAudio at all. You can try to create an ALSA loopback device (snd-aloop), configure JACK to use it as an output, and configure roc-send to use it as an input.

So the chain will be: hydrogen -> jack -> alsa loopback -> roc-send

[Here](https://gavv.github.io/articles/roc-tutorial/#run-roc-send) you can find instructions how to create the loopback device and use it with roc-send tool. You can follow them and test that everything works without JACK. Then you'll have to register this loopback device as a JACK output (I guess using `alsa_out` tool) and connect hydrogen to this JACK output.

**Comment by gavv** [comment]
I've created #254.

**Comment by jave** [comment]
I finally got around to test your suggestion to use the alsa loopback device.  I used qjacktctl to connect jackd with the alsa device. This worked well, thanks!

**Comment by ejeschke** [comment]
Stumbled across this wishing to do the same thing: stream audio out of JACK on server to simple network clients using pulse.  FWIW, I managed to get something working somewhat reasonably (time will tell) with:
```bash
jack-stdout -e float <jack-port-out1> <jack-port-out2> \
| sox -t raw -r 44100 -e float -b 32 -c 2 - -t wav - \
| roc-send -vv -s rtp+rs8m:<recv-ip>:10001 -r rs8m:<recv-ip>:10002 -d wav -i -
```
I'm wondering if I can get rid of the middle `sox` conversion to WAV (looks like `roc-send` is linking to the sox library to do the input conversions anyway).  Is there a "raw" `stdin` option to `roc-send` where I can specify the rate (e.g. 48000), bit depth (e.g. 16, 32), encoding (e.g. float, signed int), and number of channels?  Seems like if you are using `libsox` to do the conversions it would already have the machinery in place to do this...

**Comment by gavv** [comment]
@ejeschke

Thanks for sharing!

> Is there a "raw" stdin option to roc-send where I can specify the rate (e.g. 48000), bit depth (e.g. 16, 32), encoding (e.g. float, signed int), and number of channels?

There is an option to force specific sample rate (`--rate`). It's also possible to force specific file/sample format, e.g. `s16` for signed 16-bit integers (`--driver` in master, `--input-format` in develop). In develop, you can also list supported formats using `--list-supported` options). These formats are actually the same as supported by sox (you can use them instead of `raw` in `-t` option).

However, so far there is no option to force specific number of channels. I think I'll add it in 0.3.
